File: java/src/main/java/com/google/flatbuffers/StringVector.java
/*
 * Copyright 2019 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.flatbuffers;

import static com.google.flatbuffers.Constants.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.Charset;

/**
 * Helper type for accessing vector of String.
 */
public final class StringVector extends BaseVector {
  private Utf8 utf8 = Utf8.getDefault();

  /**
   * Assigns vector access object to vector data.
   *
   * @param _vector Start data of a vector.
   * @param _element_size Size of a vector element.
   * @param _bb Table's ByteBuffer.
   * @return Returns current vector access object assigned to vector data whose offset is stored at
   *         `vector`.
   */
  public StringVector __assign(int _vector, int _element_size, ByteBuffer _bb) {
    __reset(_vector, _element_size, _bb); return this;
  }

  /**
   * Reads the String at the given index.
   *
   * @param j The index from which the String value will be read.
   * @return the String at the given index.
   */
  public String get(int j) {
    return Table.__string(__element(j), bb, utf8);
  }
}

File: net/FlatSpanBuffers/Vectors/StringVector.cs
/*
 * Copyright 2026 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using Google.FlatSpanBuffers.Operations;

namespace Google.FlatSpanBuffers.Vectors
{
    public struct StringVector
    {
        private VectorData _data;

        public bool IsEmpty => _data.IsEmpty;
        public int Length => _data.Length;

        public StringVector(Table p, int offset)
        {
            _data = new VectorData(p, offset, 4);
        }

        public string this[int index]
            => TableOperations.GetString(_data.GetBufferIndex(index), _data.bb);

        public ReadOnlySpan<byte> GetBytes(int index)
            => TableOperations.GetStringBytes(_data.GetBufferIndex(index), _data.bb);

        public Enumerator GetEnumerator() => new Enumerator(ref this);

        public struct Enumerator
        {
            private ByteBuffer _bb;
            private VectorEnumeratorData _enumeratorData;

            internal Enumerator(ref StringVector vector)
            {
                _bb = vector._data.bb;
                _enumeratorData = new VectorEnumeratorData(ref vector._data);
            }

            public bool MoveNext() => _enumeratorData.MoveNext();
            public string Current => TableOperations.GetString(_enumeratorData.Current, _bb);
        }
    }

    public ref struct StringVectorSpan
    {
        private VectorSpanData _data;

        public bool IsEmpty => _data.IsEmpty;
        public int Length => _data.Length;

        public StringVectorSpan(TableSpan p, int offset)
        {
            _data = new VectorSpanData(p, offset, 4);
        }

        public string this[int index]
            => TableOperations.GetString(_data.GetBufferIndex(index), _data.bb);

        public ReadOnlySpan<byte> GetBytes(int index)
            => TableOperations.GetStringBytes(_data.GetBufferIndex(index), _data.bb);

        public Enumerator GetEnumerator() => new Enumerator(ref this);

        public ref struct Enumerator
        {
            private ByteSpanBuffer _bb;
            private VectorEnumeratorData _enumeratorData;

            internal Enumerator(scoped ref StringVectorSpan vector)
            {
                _bb = vector._data.bb;
                _enumeratorData = new VectorEnumeratorData(ref vector._data);
            }

            public bool MoveNext() => _enumeratorData.MoveNext();
            public string Current => TableOperations.GetString(_enumeratorData.Current, _bb);
        }
    }
}